sdk/src/main/scala/com/horizen/consensus/ConsensusDataProvider.scala

```diff
@@ -52,27 +57,50 @@ trait ConsensusDataProvider {
     FullConsensusEpochInfo(stakeEpochInfo, nonceEpochInfo)
   }
 
+  //Check possible stack buffer overflow situations
   def calculateNonceForEpoch(epochId: ConsensusEpochId): NonceConsensusEpochInfo = {
     val lastBlockIdInEpoch: ModifierId = lastBlockIdInEpochId(epochId)
     val lastBlockInfoInEpoch: SidechainBlockInfo = storage.blockInfoById(lastBlockIdInEpoch)
 
-    val nonceOpt: Option[BigInteger] = foldEpochRight[Option[BigInteger]](None, lastBlockIdInEpoch, lastBlockInfoInEpoch){
-      (blockId: ModifierId, blockInfo: SidechainBlockInfo, accumulator: Option[BigInteger]) =>
-        {
-          val minimalHashForCurrentBlockOpt: Option[BigInteger] = getMinimalHashOpt(blockInfo.mainchainBlockReferenceHashes.map(_.data))
-          (minimalHashForCurrentBlockOpt, accumulator) match {
-            case (None, _) => accumulator
-            case (minimalHashOpt, None) => minimalHashOpt
-            case (Some(a), Some(b)) => Option(a.min(b))
-        }
-      }
+    if (isGenesisBlock(lastBlockIdInEpoch)) {
+      ConsensusDataProvider.calculateNonceForGenesisBlockInfo(lastBlockInfoInEpoch)
     }
+    else {
+      calculateNonceForNonGenesisEpoch(lastBlockIdInEpoch, lastBlockInfoInEpoch)
+    }
+  }
 
-    require(nonceOpt.isDefined, "No mainchain reference had been found for whole consensus epoch") //crash whole world here?
-
-    NonceConsensusEpochInfo(bigIntToConsensusNonce(nonceOpt.get))
+  private def calculateNonceForNonGenesisEpoch(lastBlockIdInEpoch: ModifierId, lastBlockInfoInEpoch: SidechainBlockInfo): NonceConsensusEpochInfo = {
+    val previousEpoch = getPreviousConsensusEpochIdForBlock(lastBlockIdInEpoch, storage.blockInfoById(lastBlockIdInEpoch))
```

Comment:
can we use `lastBlockInfoInEpoch` instead of `storage.blockInfoById(lastBlockIdInEpoch)`?

Reply:
yes, fixed